### ISSUE TITLE
Permit testing error handling

### ISSUE THREAD

**Issue by johkra** [dbrgn/embedded-hal-mock#9]
Thanks for this crate! Before discovering it I had written my own mock I2C implementation which is less well designed.

Unfortunately, I could not test error handling with the I2C mock in this crate.

Extract from the code I'd like to test:

```rust
#[derive(Debug, PartialEq)]
pub enum Error<E> {
    I2c(E),
}
pub struct Si7021<I2C> {
    i2c: I2C,
}
impl<E, I2C> Si7021<I2C>
where
    I2C: i2c::WriteRead<Error = E> + i2c::Write<Error = E>,
{
    pub fn firmware_revision(&mut self) -> Result<u8, Error<E>> {
        let mut buffer = [0u8; 1];
        self.i2c.write_read(0x40, &[0x84, 0xb8], &mut buffer).map_err(Error::I2c)?;
        Ok(buffer[0])
    }
}
```

And the related test:

```rust
    #[test]
    fn get_firmware_revision_i2c_error() {
        let mut si7021 = Si7021::new(MockI2c { /* Returns NoResponse error for undefined commands */ });
        let firmware_revision = si7021.firmware_revision();
        assert!(firmware_revision.is_err());
        assert_eq!(
            firmware_revision.unwrap_err(),
            si7021_hal::Error::I2c(Error::NoResponse)
        );
    }
```
*Note: My goal is to test that there was an si7021_hal::Error::I2c error. I might be able to use the assert_matches crate to remove the PartialEq requirement resulting from using assert_eq!.*

I tried to create a similar test with the I2C mock from this crate, but encountered two issues:

1. MockError doesn't implement PartialEq. This may be because it defines an (unused?) Io error variant and io::Error doesn't implement PartialEq.
1. I couldn't find a way to make the mock return an error. If I defined no I2cTransactions or wrong transactions it will panic.

For my use case, being able to put the mock in a mode where it will always return errors would be sufficient and potentially easier than allowing specific I2C transactions to return errors.

There may be a hypothetical test case were one firsts resets the device and then polls the device (with the first tries returning an error from the underlying I2C implementation) instead of waiting a fixed amount of time, but otherwise I couldn't think of test cases where fine-grained control over errors from the mock implementation would be required.

**Comment by dbrgn** [comment]
Ok, so if I understood this correctly, you are looking for two things:

- `MockError` should implement `PartialEq`
- You want a way to mock I²C read/write errors

Is this correct?

**Comment by dbrgn** [comment]
@ryankurte what do you think about this? Do you think we could add a `I2cTransaction::error(E)` where `E` is the associated error type?

**Comment by ryankurte** [comment]
`I2cTransaction::error(E)` sounds good to me!

**Comment by darkwater** [comment]
Something similar for other components would be useful too. I want to mock a serial device, but I want to be able to call `.read()` while "no data is available". Like a `MaybeRead` transaction maybe? It shouldn't panic at least.

**Comment by dbrgn** [comment]
For the record, `PartialEq` was added to `MockError` a few months ago in 7302f701, and the `Pin` mocks already support error transactions.

I'm currently working on bringing those error transactions to `I2c` as well.

**Comment by dbrgn** [comment]
Implemented in #29.
